### PR TITLE
Use ODBC adapter

### PR DESCRIPTION
**Pull request by michaljurecko**
Slack: https://keboola.slack.com/archives/CQ1ASK06M/p1674654575328409

### CODE REVIEW

File: tests/functional/incremental-fetching-date-with-state/expected/data/out/tables/in.c-main.auto-increment-timestamp.csv.manifest [hunk 1]
```diff
@@ -36,10 +36,6 @@
                 "key": "KBC.datatype.basetype",
                 "value": "INTEGER"
             },
-            {
-                "key": "KBC.datatype.length",
-                "value": "4"
-            },
```
Comment (michaljurecko) [inline tests/functional/incremental-fetching-date-with-state/expected/data/out/tables/in.c-main.auto-increment-timestamp.csv.manifest:42]: Niekde to povodne generovalo nezmyselnu `length` pre int.

File: tests/functional/incremental-fetching-date-with-state/expected/data/out/tables/in.c-main.auto-increment-timestamp.csv.manifest [hunk 2]
```diff
@@ -62,7 +58,7 @@
             },
             {
                 "key": "KBC.ordinalPosition",
-                "value": 0
+                "value": 1
```
Comment (michaljurecko) [inline tests/functional/incremental-fetching-date-with-state/expected/data/out/tables/in.c-main.auto-increment-timestamp.csv.manifest:61]: `ordinalPosition` je cislovana od `1`, tak to ma aj `db-ex-mysql`, takze by to malo byt spravne.

File: tests/functional/incremental-fetching-date-with-state/expected/data/out/tables/in.c-main.auto-increment-timestamp.csv.manifest [hunk 3]
```diff
@@ -100,11 +96,11 @@
             },
             {
                 "key": "KBC.uniqueKey",
-                "value": true
+                "value": false
```
Comment (michaljurecko) [inline tests/functional/incremental-fetching-date-with-state/expected/data/out/tables/in.c-main.auto-increment-timestamp.csv.manifest:99]: Toto odbc nereportuje, ale neviem o tom, ze by sa to niekde v platforme pouzivalo.

**Review by michaljurecko — COMMENTED** [review]
Odporucam pozriet po commitoch

**Comment by michaljurecko** [comment]
Dal som to nad PHP 8, uvidime ci aj tu bude problem alebo sa jednalo iba o PDO.

**Review by ondrajodas — DISMISSED** [review]
Jj vypadá to dobře ale teda bylo to víc jak 30minut :)

**Comment by michaljurecko** [comment]
Jj tie PKs a testy ma zdrzali :)

File: Dockerfile [hunk 4]
```diff
@@ -5,7 +5,11 @@ ARG AWS_SESSION_TOKEN
 
 RUN /usr/bin/aws s3 cp s3://keboola-drivers/db2-odbc/v11.5.4_linuxx64_dsdriver.tar.gz /tmp/dsdriver.tar.gz
 
-FROM php:8-cli
+# Note 2023-02-03:
+# The ODBC driver doesn't work properly with PHP >= 8.2.
+# This is probably due to changes in escaping, see: https://github.com/php/php-src/issues/8300 (PHP 8.2).
+# A password that contains "@" does not work, even if it is correct.
+FROM php:8.1-cli
```
Comment (michaljurecko) [inline Dockerfile:12]: Na produkcii to neslo :cry: 

Funguje to s PHP `8.1`, nefunguje s `8.2`.

Tipujem to na toto:
- https://www.php.net/ChangeLog-8.php
- https://github.com/php/php-src/issues/8300
- 
image

Ak je v hesle `@`, tak validne heslo s `8.2` neprejde, asi sa driver spolieha, ze to nebude escapovane, alebo take nieco.

-------------

Stacilo by teda v `main` branch upravit base tag.
No tato verzia s ODBC adapterom ma fixnute aj dalsie drobnosti co som komentoval a nema ziaden specialny kod (na rozdiel od PDO implementacie, co tu bola predtym), tj. malo by byt lepsie pouzit toto.

Comment (michaljurecko) [reply Dockerfile:12]: Testovaci job uz ide (resp. spadne spravne na tom, ze tabulka neexistuje).

Job: https://connection.eu-central-1.keboola.com/admin/projects/3666/queue/528117588

image

:heavy_check_mark: connection works
image

**Review by ondrajodas — APPROVED** [review]

File: Dockerfile (hunk 4 above)
Comment (NattyNarwhal) [reply Dockerfile:12]: Hi, I'm the person who developed the escaping feature for the ODBC extensions - I found this PR linked back from the original issue on PHP upstream. If there's a regression, I'd like to know, or have it filed as a bug upstream. I don't speak Czech, so I don't have much context to go on here.

Comment (michaljurecko) [reply Dockerfile:12]: Hi @NattyNarwhal , Thanks for your interest.

The situation was as follows:
- We have updated the Dockerfile to `PHP 8.2`.
- A client reported that his connection stopped working.
- We did not notice the problem with our test DB (see `docker-compose.yml`).
- We found that this applies to both direct use of `ODBC` and `PDO` + `ODBC`.
- When we downgraded the Dockerfile to `PHP 8.1`, it started working again.
- We noticed that the client has an `@` character in the password.
- The GH-8300 is the only thing we found in the PHP `8.1 vs 8.2` changelog that could match this.
- We did not investigate further.

Comment (michaljurecko) [reply Dockerfile:12]: The connection is created as follows:
- https://github.com/keboola/db-extractor-db2/blob/2a9903613721708097194d3ef6cdbb99705cc503/src/Extractor/DB2OdbcConnection.php#L14-L29
- parent https://github.com/keboola/db-extractor-adapter/blob/7cfa4d27025d2874a9644ca21eda26a38723d753/src/ODBC/OdbcConnection.php#L60-L69

Comment (michaljurecko) [reply Dockerfile:12]: I see, there is no `@` quoting
https://github.com/php/php-src/blob/2920a266364aef84df82cbb2b01b573705fa57ec/main/php_odbc_utils.c#L93

Maybe is the DB2 driver incompatible with the quoting format?
But why would it work for us in CI tests and not for the client? I do not know.

Comment (NattyNarwhal) [reply Dockerfile:12]: We use the Db2 for i driver and haven't had any issues (in fact, the quoting thing was to solve issues of using i.e. `{` and `}` within a password. I don't know about the Db2 LUW driver.

A `@` would [result in it being quoted](https://github.com/php/php-src/blob/master/main/php_odbc_utils.c#L72); I don't think there's any other special handling for that character?

If you manually quote the password (i.e. `{p@ssword}`), does that work?

Comment (michaljurecko) [reply Dockerfile:12]: Unfortunately, I no longer have access to the client database.

I can try to replicate it against our test DB with `@` in the password, but I can't promise when I'll have time, probably no sooner than after 14 days.

Comment (NattyNarwhal) [reply Dockerfile:12]: I'll try to reproduce with other drivers in the meantime then.

**Pull request merged**